### ISSUE TITLE
[Request] Selectable lang tags

### ISSUE THREAD

**Issue by lars18th** [stari4ek/tvirl#39]
Hi @stari4ek ,

I have one trouble that you can fix without much effort. The problem is that my EPG source uses more than one language. Then some entries have multiple entries with different `lang=` tag. In the example of the Wiki some examples exists:
https://github.com/stari4ek/tvirl/wiki/M3U#matching-m3u-channel-to-epg

Then I want to select the language to incorporate the metadata. For example with a single tag in the M3U list like this `tvg-lang` (used by some other tools):
```
#EXTINF:0 tvg-id="493" tvg-lang="en" tvg-logo="http://best-logo.net/images/euronews.png" group-title="NEWS",Euronews
udp://@235.1.1.2:5678
```
So, when reading the XMLTV, if some entry is in multiple languages then only the one matching with the `tvg-lang` will be used. And if only one entry exists, then use this even if it not maches.

What you think? You can incorporate this extension to fix multi-language EPG sources?
Thank you!

**Comment by stari4ek** [comment]
Sounds legit.
Do you have any other references (docs, code samples, other projects) for `tvg-lang` description?

**Comment by lars18th** [comment]
Hi @stari4ek ,

> Do you have any other references (docs, code samples, other projects) for `tvg-lang` description?

As you know (I supose) the M3U format is very Community driven. So, almost I see this two alternatives:
- `tvg-lang=` with values like `en` or `eng` (two or three letters). This indicates more or less the preference of language for audio when multiple tracks are found in the source. The SSIPTV software uses this tag in this way. My suggestion is to use this tag and do the comparision with exact match (so, don't need to identify the language).
- `tvg-language=` with values in ISO format like `tvg-language="English"`. In this case, the tag is not indicating any user selection but a list of audio tracks in the source. So the tag represents a list of languages. This is the same way as the tag `tvg-country` works that appears in severals lists (for grouping). One documentation referencing this can be found at https://gitee.com/jjm2473/iptv/blob/master/.github/CONTRIBUTING.md

I hope you want to implement this.
Regards.

**Comment by lars18th** [comment]
Hi @stari4ek ,

I have one "simple" suggestion until you can implement a full solution. The first step can be to filter the XMLTV entries using the current user language. So, then it will be a "global" user selection. But almost it will be preferable to the current behaviour that in fact is use the _last_ entry in the XML. See this example (from https://github.com/googlesamples/androidtv-sample-inputs/blob/master/xmltv.dtd):

```
  <programme start="200006031633" channel="3sat.de">
    <title lang="de">blah</title>
    <title lang="en">blah</title>
...
```
At time the `title` used is the one in English ("en") only because is the last readed. Then when interchanging the lines the version used is the German one ("de"). So my suggestion is to first implement a simple filter:
- Inside a programme for each tag with a `lang=` parameter, discard all that doesn't match the user language.

I hope you want to implement (soon) this.
Thank you!

**Comment by stari4ek** [comment]
Current implementation honors user's locale and prefers it if available.
if it works differently for you please let me know with real example

**Comment by lars18th** [comment]
Hi @stari4ek ,

> Current implementation honors user's locale and prefers it if available.

Current == Which version? I'm subscribed to BETA.
User's locale == Android TV configured language?
Prefers == Inside tvirl?

Please, if you can describe it more clearly then I'll try if I can fix the trouble.

> if it works differently for you please let me know with real example

In my current case the source has two `lang=` entries for each value of `title`, `sub-title`, `desc` and `category` tags. And in the guide I see the values of the _second_ value.

**Comment by stari4ek** [comment]
> Current == Which version? I'm subscribed to BETA.
> User's locale == Android TV configured language?
> Prefers == Inside tvirl?

All recent versions - there was no changes recently for this functionality.
user's locale is language preferences configured in system settings (Android TV)

If there are multiple strings associated with title/description - TVirl selects one of it depending on it's own logic (user's locale).

Please provide exact example (xmltv file) with the name of problematic channel and language selected in system settings

**Comment by davidpolh** [comment]
I have this problem as well. My EPG source is https://github.com/MPAndrew/EpgGratis/raw/master/guide.xml.gz and my Andoid locale is "es_ES" [ Spanish (Spain) ]. Until now I had not understood why I see many entries in English in a lot of channels.

**Comment by lars18th** [comment]
Hi @stari4ek 

> Please provide exact example (xmltv file) with the name of problematic channel and language selected in system settings

It's quite simple to check: Take _any_ XMLTV file and edit it duplicating the title entry like this:
```
  <programme start= ... >
    <title lang="de">DE-TITLE</title>
    <title lang="en">EN-TITLE</title>
...
```
and check what you see in the EPG if you select the "de_AT" locale for example. I'm sure you will see "EN-TITLE" and not the "DE-TITLE".

**Comment by stari4ek** [comment]
@davidpolh, thank you for providing specific EPG to investigate.
I will check how exactly it's processed by TVirl, but here are some issues I see with it:
```xml
<programme start="20210218024100 +0100" stop="20210218032500 +0100" channel="CANAL ODISEA">
    <title lang="es">[COLOR yellow]Los hechos más extraños de las Guerras Mundiales[/COLOR]</title>
    <title lang="en">Los hechos más extraños de las Guerras Mundiales</title>
    <title lang="en">Los hechos más extraños de las Guerras Mundiales</title>
    <sub-title lang="es">[COLOR orange]S3 E8 [/COLOR]</sub-title>
    <sub-title lang="en">S3 E8  </sub-title>
    <desc lang="es">[B]2018[/B] / [B]Edad:[/B] [B][COLOR yellow]12[/COLOR][/B] / [COLOR yellow]★★★★★[/COLOR] 
La fuga del campo de concentración de Sobibor se tuerce en el último momento, y el gobierno prefiere aceptar una teoría OVNI antes que la verdad.
[B][COLOR orange]Director:[/COLOR][/B] Mike Ibeji. </desc>
    <desc lang="en">2018 | Edad: 12 | ★★★★★ 
La fuga del campo de concentración de Sobibor se tuerce en el último momento, y el gobierno prefiere aceptar una teoría OVNI antes que la verdad.
Director: Mike Ibeji. </desc>
    <credits>
      <director>Mike Ibeji</director>
    </credits>
    <date>2018</date>
    <category lang="es">Documentary</category>
    <category lang="es">Historia</category>
    <icon src="https://www.movistarplus.es/recorte/n/caratula4/F3873825" />
    <country lang="es">R.U</country>
    <episode-num system="onscreen">S3 E8</episode-num>
    <rating system="ES">
      <value>12</value>
      <icon src="http://www.plus.es/assets/images/movistar-plus/calf-12.svg" />
    </rating>
    <star-rating system="ES">
      <value>★★★★★</value>
    </star-rating>
  </programme>
```
* multiple EN entries with same content
* ES variant has some kind of embedded formatting tags, which are not supported by TVirl/`TV Input Framework`, so if TVirl starts using it as-is, you will see crap inside title/description.
* TVirl can pre-process text to drop it, but it will make XmlTv processing slower for 99.99% of users
* It looks like there is another version of XmlTv without those tags ( https://github.com/MPAndrew/EpgGratis/raw/master/epgsinrex.xml.gz)
* And, btw, some strings are just not translated:
```xml
<programme start="20210214154500 +0100" stop="20210214163000 +0100" channel="DIVINITY">
    <title lang="es">[COLOR yellow]The Good Doctor[/COLOR]</title>
    <title lang="en">The Good Doctor</title>
    <title lang="en">The Good Doctor</title>
```
so on top of issue in TVirl with locale honoring there is an issue with EPG itself (it's partially translated)

**Comment by stari4ek** [comment]
@davidpolh, on top of that there is some inconsistency:
* EN version has better title
```xml
<title lang="es">Energía</title>
<title lang="en">Energía: el motor de la historia</title>
```
* ES is localized (better)
```xml
<title lang="es">Cómo entrenar a tu dragón 2</title>
<title lang="en">How to Train Your Dragon 2</title>
```
* Both variants are localized:
```xml
 <title lang="es">Los crímenes de Pembrokeshire</title>
 <title lang="en">Los crímenes de Pembrokeshire</title>
```

I have identified the issue with locale selection. Will work on it for the next release.

**Comment by lars18th** [comment]
Hi @stari4ek ,

> I have identified the issue with locale selection. Will work on it for the next release.

Great to hear that you fixed the bug with the locale settings! I can't wait to try the beta (I'm subscribed to it). :+1:

**Comment by stari4ek** [comment]
fixed for 0.5.3.1
